modules/renter/contractor/contracts.go

```diff
@@ -0,0 +1,395 @@
+			if err != nil {
+				c.log.Printf("WARN: failed to renew contract %v with %v: %v\n", id, oldContract.NetAddress, err)
+				return
+			}
+			c.log.Printf("Renewed contract %v with %v\n", id, oldContract.NetAddress)
+
+			// Lock the contractor as we update it to use the new contract
+			// instead of the old contract.
+			c.mu.Lock()
+			defer c.mu.Unlock()
+
+			// Store the contract in the record of historic contracts.
+			_, exists := c.contracts[oldContract.ID]
+			if exists {
+				c.oldContracts[oldContract.ID] = oldContract
+				delete(c.contracts, oldContract.ID)
+			}
+
+			// Add the new contract, including a mapping from the old
+			// contract to the new contract.
+			c.contracts[newContract.ID] = newContract
+			c.renewedIDs[oldContract.ID] = newContract.ID
+			c.cachedRevisions[newContract.ID] = c.cachedRevisions[oldContract.ID]
+			delete(c.cachedRevisions, oldContract.ID)
+
+			// Save the contractor.
+			err = c.saveSync()
+			if err != nil {
+				c.log.Println("Failed to save the contractor after creating a new contract.")
+			}
+		}()
+
+		// Soft sleep for a minute to allow all of the transactions to propagate
+		// the network.
+		select {
+		case <-c.tg.StopChan():
+			return
+		case <-time.After(contractFormationInterval):
+		}
+	}
+
+	// Count the number of contracts, and add more where they are needed.
+	//
+	// TODO: Use the IsGoodForUpload logic.
+	c.mu.Lock()
+	neededContracts := int(c.allowance.Hosts) - len(c.onlineContracts())
+	c.mu.Unlock()
+	// Nothing to do if there are no needed contracts.
+	if neededContracts < 1 {
+		return
+	}
+
+	// Assemble an exclusion list that includes all of the hosts that we already
+	// have contracts with, then select a new batch of hosts to attempt contract
+	// formation with.
+	c.mu.RLock()
+	var exclude []types.SiaPublicKey
+	for _, contract := range c.contracts {
+		exclude = append(exclude, contract.HostPublicKey)
+	}
+	c.mu.RUnlock()
+	hosts := c.hdb.RandomHosts(neededContracts*2+10, exclude)
+
+	// Form contracts with the hosts one at a time, until we have enough
+	// contracts.
+	for _, host := range hosts {
+		// Quit in the event of shutdown.
+		select {
+		case <-c.tg.StopChan():
+			return
+		default:
+		}
+
+		// Attempt forming a contract with htis host.
+		newContract, err := c.managedNewContract(host, numSectors, endHeight)
+		if err != nil {
+			c.log.Printf("Attempted to form a contract with %v, but if failed: %v\n", host.NetAddress, err)
+			continue
+		}
+
+		// Add this contract to the contractor and save.
+		c.mu.Lock()
+		c.contracts[newContract.ID] = newContract
+		err = c.saveSync()
+		c.mu.Unlock()
+		if err != nil {
+			c.log.Println("Unable to save the contractor:", err)
+		}
+
+		// Quit the loop if we've replaced all needed contracts.
+		neededContracts--
+		if neededContracts < 1 {
+			break
+		}
+
+		// Soft sleep before making the next contract.
+		select {
+		case <-c.tg.StopChan():
+			return
```

Comment:
I am in agreement. Perhaps in another PR, perhaps the same one where we drop the `Flush` function and transit to the NebulousLabs/threadgroup package.